pkg/endpointmanager/manager.go

```diff
@@ -219,6 +213,61 @@ func Remove(ep *endpoint.Endpoint) {
 			log.WithError(err).WithField("state", state).Warning("Unable to release endpoint ID")
 		}
 	}
+}
+
+// RuleCacheOwner is an interface representing anything which needs to be
+// notified upon endpoint removal from the endpointmanager.
+type RuleCacheOwner interface {
+	// ClearPolicyConsumers removes references to the specified id from the
+	// policy rules managed by RuleCacheOwner.
+	ClearPolicyConsumers(id uint16) *sync.WaitGroup
+}
+
+// WaitEndpointRemoved waits until all operations associated with Remove of
+// the endpoint have been completed.
+func WaitEndpointRemoved(ep *endpoint.Endpoint, owner RuleCacheOwner) {
+	select {
+	case <-Remove(ep, owner):
+		return
+	}
+}
+
+// Remove removes the endpoint from the global maps and releases the node-local
+// ID allocated for the endpoint.
+// Must be called with ep.Mutex.RLock held. Releasing of the ID of the endpoint
+// is done asynchronously. Once the ID of the endpoint is released, the returned
+// channel is closed.
+func Remove(ep *endpoint.Endpoint, owner RuleCacheOwner) <-chan struct{} {
+
+	epRemoved := make(chan struct{})
+
+	// Since the endpoint is being deleted, we no longer need to run events
+	// in its event queue.
+	ep.CloseEventQueue()
```

Comment:
I pushed it up before the endpoint is removed from the `endpointmanger`. The queue is  also closed called here in case the caller for some reason does not call it, in which case it is effectively a no-op.